File: packages/react-native-fast-io/ios/HybridInputStream.swift
//
//  HybridInputStream.swift
//  FastIO
//
//  Created by Mike Grabowski on 06/11/2024.
//

import Foundation
import NitroModules

class HybridInputStream : HybridInputStreamSpec {
  let stream : InputStream
  
  init(stream: InputStream) {
    self.stream = stream
  }
  
  func open() throws -> Void {
    stream.open()
  }
  
  func read() throws -> Promise<ArrayBufferHolder> {
    let promise = Promise<ArrayBufferHolder>()

    Task {
      let size = Int(HybridStreamFactory.BUFFER_SIZE)
      let data = UnsafeMutablePointer<UInt8>.allocate(capacity: size)
      
      let bytesRead = stream.read(data, maxLength: size)
      
      let deleteFunc = {
        data.deallocate()
      }
      
      if (bytesRead >= 0) {
        promise.resolve(withResult: ArrayBufferHolder.wrap(dataWithoutCopy: data, size: bytesRead, onDelete: deleteFunc))
      } else {
        deleteFunc()
        promise.reject(withError: stream.streamError ?? RuntimeError.error(withMessage: "Unexpected error reading stream"))
      }
    }
    
    return promise
  }
  
  func close() {
    stream.close()
  }
  
  var hybridContext = margelo.nitro.HybridContext()
  
  // Return size of the instance to inform JS GC about memory pressure
  var memorySize: Int {
    return getSizeOf(self)
  }
}

File: packages/react-native-fast-io/android/src/main/java/com/margelo/nitro/fastio/HybridInputStream.kt
package com.margelo.nitro.fastio

import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.InputStream

class HybridInputStream(val stream: InputStream) : HybridInputStreamSpec() {
    override fun read(): Promise<ArrayBuffer> {
        return Promise.async {
            withContext(Dispatchers.IO) {
                val bytes = ByteArray(HybridStreamFactory.BUFFER_SIZE)
                val bytesRead = stream.read(bytes, 0, bytes.size)

                when {
                    bytesRead == -1 -> {
                        val emptyBuffer = ArrayBuffer.allocate(0)
                        emptyBuffer
                    }

                    bytesRead > 0 -> {
                        val arrayBuffer = ArrayBuffer.allocate(bytesRead)

                        val destBuffer = arrayBuffer.getBuffer(false)
                        destBuffer.put(bytes, 0, bytesRead)

                        arrayBuffer
                    }

                    else -> {
                        throw Error("Unexpected error reading stream")
                    }
                }
            }
        }
    }

    override fun open() {
        // No explicit open needed for Java InputStreams
    }

    override fun close() {
        stream.close()
    }

    override val memorySize: Long
        get() = try {
            stream.available().toLong()
        } catch (e: Exception) {
            0L
        }
} 
